src/relay/backend/contrib/example_target_hooks/relay_to_tir.cc

```diff
@@ -0,0 +1,131 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/transform.h>
+#include <tvm/tir/buffer.h>
+#include <tvm/tir/builtin.h>
+#include <tvm/tir/expr.h>
+#include <tvm/tir/function.h>
+#include <tvm/tir/op.h>
+
+namespace tvm {
+namespace relay {
+namespace contrib {
+namespace example_target_hooks {
+
+class ConvertAddToSubtract : public MixedModeMutator {
+ public:
+  explicit ConvertAddToSubtract(IRModule ir_module, Target host_target)
+      : ir_module_(ir_module), host_target_(host_target) {}
+
+  IRModule Mutate() {
+    GlobalVar main_global_var = ir_module_->GetGlobalVar("main");
+    BaseFunc main = ir_module_->Lookup(main_global_var);
+    Function main_func = GetRef<Function>(main.as<FunctionNode>());
+
+    // Copy everything across and mutate the body
+    Function mutated_main =
+        Function(main_func->params, VisitExpr(main_func->body), main_func->ret_type,
+                 main_func->type_params, main_func->attrs, main_func->span);
+
+    ir_module_->Update(main_global_var, mutated_main);
+
+    return ir_module_;
+  }
+
+ private:
+  tir::Load LoadIndex(const tir::Buffer& buffer, const PrimExpr& index) {
+    return tir::Load(DataType::Float(32), buffer->data, index, tir::const_true());
+  }
+
+  void ReplaceAddWithSubtractPrimFunc(const GlobalVar& new_global_var, const Function& func) {
+    tir::Buffer x_buffer = tir::decl_buffer({8}, DataType::Float(32), "x");
+    tir::Buffer y_buffer = tir::decl_buffer({8}, DataType::Float(32), "y");
+    tir::Buffer out_buffer = tir::decl_buffer({8}, DataType::Float(32));
+
+    tir::Var x_var("x", DataType::Handle());
+    tir::Var y_var("y", DataType::Handle());
+    tir::Var out_var("out", DataType::Handle());
+
+    Map<String, ObjectRef> dict_attrs;
+    dict_attrs.Set("global_symbol", new_global_var->name_hint);
+    dict_attrs.Set("tir.noalias", Bool(true));
+
+    te::Var index("index", DataType::Int(32));
+    tir::Sub indexed_sub = tir::Sub(LoadIndex(x_buffer, index), LoadIndex(y_buffer, index));
+    tir::Stmt math_body = tir::Store(out_buffer->data, indexed_sub, index, tir::const_true());
+    tir::Stmt math_loop = tir::For(index, 0, 8, tir::ForKind::kSerial, math_body);
+
+    Map<tir::Var, tir::Buffer> buffer_map = {
+        {x_var, x_buffer},
+        {y_var, y_buffer},
+        {out_var, out_buffer},
+    };
+
+    tir::PrimFunc replacement_func = tir::PrimFunc({x_var, y_var, out_var}, math_loop, VoidType(),
```

Comment:
It all looks pretty slick to me, you've convinced me of the 'it's just a Pass' approach. The only part not demoed here I can think of is caching. Would you be up for adding that too? Then I think we shouldn't get hung up on trying to fold any of this handling back into the te_compiler (eg by inheriting from some 'GenericLowerTE' class or something). Thanks.

Reply:
My first thought is the `Pass` infra works well as a composition approach where we give a series of tools to the user rather than having them extend from specific classes - to incorporate caching I'd change the signature to something like `tvm::transform::Pass(const CompileCache& cache)` or similar so we can pass the cache between the intermediary passes? 

Splitting the cache from `te_compiler.cc` and factoring it in like that seems big enough for a separate PR?

Reply:
Agree on reuse by composition rather than inheritance. I wasn't proposing to refactor caching to be shared, rather proposing to show it in your example of something that's so easy to do directly we shouldn't even worry about refactoring.

After writing my initial comment I remembered the prim<->prim shape function handling is also a bit subtle, but that's quite possibly something this sort of extension mechanism won't need to support anyway. Or, by the time it does we will have cleaned that part up enough the way to handle it will be obvious.

Reply:
The reason I'd suggest using a shared cache passed between the passes is that it encapsulates the logic of:
```
(Function, Target) -> PrimFunc / External Function / Empty Node
```
Which includes things like tracking the `UniqueName` of a node, we could use the `IRModule` itself as a form of cache in the `relay_to_tir.cc` but it'd likely be better to give the full cache capability to the hooks so we don't have to re-implement too much of that.

That's what I think justifies factoring the cache out and using it between the constituent passes rather than a local cache per `Pass`. What do you think?

Reply:
I don't see it's necessary.
Within a pass it's a few lines to ensure two calls to the same attrs::kPrimitive Function are rewritten to call the same PrimFunc. We could probably even use the memoization built into ExprMutator.
Between passes there's nothing left to say -- it has all been encapsulated within the IRModule itself.
Even if it were necessary for some peculiar reason then I'd turn the conversation to figuring out how to extend IRModule or attributes or whatever to again ensure there's no special state between passes other than what is spelled out in the IRModule.
Does that make sense?

BTW LGTM for this one exactly as is since I can see the caching issue has deserved more conversation who's outcome can easily go into a follow up. Thanks for pushing on the 'just a Pass' approach, it's so much better :-)